### ISSUE TITLE
transformer encoder-layer, the sample-Independent attn_mask(dim=3) has different behaviors when training and validating

### ISSUE THREAD

**Issue by liujiaozhiren** [pytorch/pytorch#102333]
### 🐛 Describe the bug

in torch.nn.modules.transormer
line 469  TransformerEncoderLayer.forward()
it supports attn_mask with dim=3, whose first dim represent diff sample in batch, 
however in this function(at line 543), it invokes the attention.merge_masks()  which one support the attn_mask with dim=2 and do nothing for the raw attn_mask with dim=3.

the code logic is so simple so there is no use case, please start with torch.nn.modules.transormer line 469 to line 543

I just simply check the latest version, Probably the bug still exists.

### Versions

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.25.1
Libc version: N/A

Python version: 3.9.16 (main, Mar  8 2023, 04:29:24)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchvision==0.15.0
[conda] numpy                     1.23.5           py39h1398885_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy-base                1.23.5           py39h90707a3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] pytorch                   2.0.0                   py3.9_0    pytorch
[conda] torchaudio                2.0.0                  py39_cpu    pytorch
[conda] torchvision               0.15.0                 py39_cpu    pytorch
